### PR TITLE
Add withdrawWithSignature function

### PR DESCRIPTION
**Pull request by francisldn**
#747 add withdrawWithSignature function. Edited IPool.sol, Pool.sol and unit.js files.

### CODE REVIEW

File: source/pool/contracts/Pool.sol [hunk 1]
```diff
@@ -258,6 +263,38 @@ contract Pool is IPool, Ownable {
...
+      address signer = ECDSA.recover(ethSignedMessageHash, v,r,s);
+      require(admins[signer], "NOT_VERIFIED");
+      // mark the hash of (signature, hash and nonce) and recipient as true
+      hashClaimed[calculatedHash] = true;
+      IERC20(token).safeTransfer(msg.sender, amount);
+      emit WithdrawWithSignature(signer, token, amount, msg.sender, nonce);
+      return amount;
+    }
+
+
```
Comment (AboldUSER) [inline source/pool/contracts/Pool.sol:297]: There is an extra empty line here that can be deleted

File: source/pool/contracts/Pool.sol [hunk 2]
```diff
@@ -363,4 +400,7 @@ contract Pool is IPool, Ownable {
     bytes32 leaf = keccak256(abi.encodePacked(participant, score));
     return MerkleProof.verify(proof, root, leaf);
   }
-}
+
+
+
```
Comment (AboldUSER) [inline source/pool/contracts/Pool.sol:405]: Three empty lines that can be deleted

File: source/pool/contracts/interfaces/IPool.sol [hunk 3]
```diff
@@ -83,4 +90,14 @@ interface IPool {
     uint256 score,
     bytes32[] memory proof
   ) external pure returns (bool valid);
-}
+
+
+  function withdrawWithSignature(
```
Comment (AboldUSER) [inline source/pool/contracts/interfaces/IPool.sol:95]: I suggest moving the function into the same order as in the actual contract

File: source/pool/test/integration.js [hunk 4]
```diff
@@ -54,13 +54,14 @@ describe('Pool Integration Tests', () => {
...
       await ethers.getContractFactory(STAKING.abi, STAKING.bytecode)
-    ).deploy(feeToken.address, 'StakedAST', 'sAST', 100, 1)
+    ).deploy(feeToken.address, 'StakedAST', 'sAST', 100, 1) 
     await stakeContract.deployed()
 
     pool = await (
       await ethers.getContractFactory('Pool')
     ).deploy(CLAIM_SCALE, CLAIM_MAX, stakeContract.address, feeToken.address)
     await pool.deployed()
+
```
Comment (AboldUSER) [inline source/pool/test/integration.js:65]: An empty line that can be deleted

File: source/pool/test/integration.js [hunk 5]
```diff
@@ -293,8 +294,9 @@ describe('Pool Integration Tests', () => {
         )
       ).to.be.revertedWith('INVALID_TOKEN')
     })
-  })
 
+
```
Comment (AboldUSER) [inline source/pool/test/integration.js:298]: An empty line that can be deleted

File: source/pool/test/unit.js [hunk 6]
```diff
@@ -456,6 +456,57 @@ describe('Pool Unit Tests', () => {
         )
       ).to.be.revertedWith('INVALID_TOKEN')
     })
+
+    it('withdrawWithSignature does not revert with signer who is admin', async () => {
```
Comment (AboldUSER) [inline source/pool/test/unit.js:460]: This can be named `withdrawWithSignature success` to align with similar withdraw function tests

**Review by AboldUSER — CHANGES_REQUESTED** [review]
Looks good, just added some formatting updates.
Also I think for 100% test coverage you'll need a test that reverts when a signature is used a second time on withdrawWithSignature

**Comment by francisldn** [comment]
Thanks for your comments @AboldUSER. I have added 1 more test to improve test coverage and tidied up the formatting issues in contracts and test files.

**Comment by francisldn** [comment]
The PR is ready for further review. I merged the changes in main, added a test in unit test file to bring the test coverage to 100% and did the formatting changes as suggested. No change to the integration test file as per the main branch.

**Review by AboldUSER — APPROVED** [review]

**Pull request merged**